cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp

```diff
@@ -2601,4 +2615,169 @@ SizeType32 TrtGptModelInflightBatching::getMaxCapacityBatchSize(SizeType32 input
+    }
+
+    SizeType32 contextId = 0;
+    for (auto const& llmReq : contextRequests)
+    {
+        if (llmReq->isFirstContextChunk() && isBeforePrepareBuffers)
+        {
+            // For first chunk: Blocking prefetch on runtime stream to ensure data is ready
+            processPromptTableChunk(llmReq, true, bufferId, contextId);
+        }
+        else if (!isBeforePrepareBuffers) // prefetching for subsequent chunks
+        {
+            // For the first prefetch chunk, don't need to wait for previous prefetch to complete
+            // For subsequent chunks: Need to wait for previous prefetch to complete
+            if (!llmReq->isFirstContextChunk())
+            {
+                mRuntime->getBufferManager().getStream().wait(mPtableCopyDoneEvent);
+            }
+
+            // Non-blocking prefetch on copy stream to prepare next chunk in pong buffer
+            if (llmReq->getContextRemainingLength() > 0)
+            {
+                processPromptTableChunk(llmReq, false, bufferId, contextId);
+            }
+        }
+
+        ++contextId;
+    }
+}
+
+void TrtGptModelInflightBatching::processPromptTableChunk(
+    std::shared_ptr<LlmRequest> const& llmReq, bool isBeforePrepareBuffers, SizeType32 bufferId, SizeType32 contextId)
+{
+    NVTX3_SCOPED_RANGE_WITH_NAME(range, "processPromptTableChunk");
+    auto& promptTuningBuffers = mBuffers[bufferId]->getPromptTuningBuffers();
+    auto const chunkSize = llmReq->getContextChunkSize();
+    auto& inputTokensMutable = llmReq->getTokensMutable(0);
+    auto vocabSize = mModelConfig.getVocabSize();
+
+    // For first chunk's initialization
+    if (isBeforePrepareBuffers)
+    {
+        promptTuningBuffers->initializeChunkPtableBuffers(mRuntime->getBufferManager(), mModelConfig, chunkSize);
+    }
+
+    size_t processChunkSize;
+    size_t beginPos;
+
+    if (!isBeforePrepareBuffers)
+    {
+        processChunkSize = std::min(chunkSize, llmReq->getContextRemainingLength() - chunkSize);
+    }
+    else
+    {
+        processChunkSize = std::min(chunkSize, llmReq->getContextRemainingLength());
+    }
+
+    if (!isBeforePrepareBuffers)
+    {
+        // For prefetching next chunk
+        if (llmReq->getContextRemainingLength() - chunkSize <= 0)
+        {
+            promptTuningBuffers->updateBufferStartPosition(promptTuningBuffers->getChunkPtableCurrentIndex(), 0);
+            return; // No more chunks to prefetch
+        }
+        beginPos = llmReq->getContextCurrentPosition() + chunkSize;
+    }
+    else
+    {
+        // For current chunk
+        beginPos = llmReq->getContextCurrentPosition();
+    }
+
+    // Bounds check
+    if (beginPos + processChunkSize > inputTokensMutable.size())
+    {
+        TLLM_THROW("Invalid chunk access: beginPos(%zu) + processChunkSize(%zu) > totalSize(%zu)", beginPos,
+            processChunkSize, inputTokensMutable.size());
+        return;
+    }
+
+    // Process tokens
```

Comment:
would be great to add a high-level explanation on the code logic here, or at the beginning of this function